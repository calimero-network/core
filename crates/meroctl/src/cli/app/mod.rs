use clap::{Parser, Subcommand};

use super::RootArgs;

mod install;
mod list;

#[derive(Parser, Debug)]
pub struct AppCommand {
    #[command(subcommand)]
    pub subcommand: AppSubCommands,
}

#[derive(Debug, Subcommand)]
pub enum AppSubCommands {
    Install(install::InstallCommand),
    #[command(alias = "ls")]
    List(list::ListCommand),
}

impl AppCommand {
    pub async fn run(self, args: RootArgs) -> eyre::Result<()> {
        match self.subcommand {
            AppSubCommands::Install(install) => install.run(args).await,
            AppSubCommands::List(list) => list.run(args).await,
        }
    }
}
