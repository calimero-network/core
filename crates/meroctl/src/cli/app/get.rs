use clap::{Parser, ValueEnum};
use eyre::{bail, Result as EyreResult};
use reqwest::Client;

use crate::cli::RootArgs;
use crate::common::RequestType::GET;
use crate::common::{get_response, multiaddr_to_url};
use crate::config_file::ConfigFile;

#[derive(Parser, Debug)]
pub struct GetCommand {
    #[arg(long, short)]
    pub method: GetValues,

    #[arg(long, short)]
    pub app_id: String,
}
#[derive(ValueEnum, Debug, Clone)]
pub enum GetValues {
    Details,
}

impl GetCommand {
    pub async fn run(self, args: RootArgs) -> EyreResult<()> {
        let path = args.home.join(&args.node_name);

        if !ConfigFile::exists(&path) {
            bail!("Config file does not exist")
        };

        let Ok(config) = ConfigFile::load(&path) else {
            bail!("Failed to load config file")
        };

        let Some(multiaddr) = config.network.server.listen.first() else {
            bail!("No address.")
        };

        let client = Client::new();

        let url = multiaddr_to_url(
            multiaddr,
            &format!("admin-api/dev/applications/{}", self.app_id),
        )?;

        let response = get_response(&client, url, None::<()>, &config.identity, GET).await?;

        if !response.status().is_success() {
            bail!("Request failed with status: {}", response.status())
        }

        println!("{}", response.text().await?);

        Ok(())
    }
}
