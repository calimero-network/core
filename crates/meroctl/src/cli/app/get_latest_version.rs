use clap::Parser;
use eyre::Result;

use crate::cli::Environment;

#[derive(Clone, Parser, Debug)]
#[command(about = "Get the latest version of a package")]
pub struct GetLatestVersionCommand {
    #[arg(help = "Package name (e.g., com.example.myapp)")]
    pub package: String,
}

impl GetLatestVersionCommand {
    pub async fn run(self, environment: &mut Environment) -> Result<()> {
        let client = environment.client()?;

        let response = client.get_latest_version(&self.package).await?;

        environment.output.write(&response);
        Ok(())
    }
}
