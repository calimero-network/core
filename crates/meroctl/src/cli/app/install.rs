use calimero_primitives::hash::Hash;
use calimero_server_primitives::admin::{
    InstallApplicationRequest, InstallApplicationResponse, InstallDevApplicationRequest,
};
use camino::Utf8PathBuf;
use clap::Parser;
use eyre::{bail, Result};
use reqwest::Client;
use semver::Version;
use tracing::info;
use url::Url;

use crate::cli::RootArgs;
use crate::common::RequestType::POST;
use crate::common::{get_response, multiaddr_to_url};
use crate::config_file::ConfigFile;

#[derive(Debug, Parser)]
pub struct InstallCommand {
    /// Path to the application
    #[arg(long, short, conflicts_with = "url")]
    pub path: Option<Utf8PathBuf>,

    /// Url of the application
    #[clap(long, short, conflicts_with = "path", requires = "metadata")]
    pub url: Option<String>,

    #[clap(short, long, help = "Version of the application")]
    pub version: Option<Version>,

    #[clap(short, long, help = "Metadata for the application")]
    pub metadata: Option<Vec<u8>>,

    #[clap(long, help = "Hash of the application")]
    pub hash: Option<Hash>,
}

impl InstallCommand {
    pub async fn run(self, args: RootArgs) -> Result<()> {
        let path = args.home.join(&args.node_name);

        if !ConfigFile::exists(&path) {
            bail!("Config file does not exist")
        };

        let Ok(config) = ConfigFile::load(&path) else {
            bail!("Failed to load config file")
        };

        let Some(multiaddr) = config.network.server.listen.first() else {
            bail!("No address.")
        };

        let client = Client::new();

        let mut is_dev_installation = false;

        let install_request = if let Some(app_path) = self.path {
            let install_dev_request = InstallDevApplicationRequest::new(
                app_path.canonicalize_utf8()?,
                self.version,
                self.metadata.unwrap_or_default(),
            );
            is_dev_installation = true;
            serde_json::to_value(install_dev_request)?
        } else if let Some(app_url) = self.url {
            let install_request = InstallApplicationRequest::new(
                Url::parse(&app_url)?,
                self.version,
                self.hash,
                self.metadata.unwrap_or_default(),
            );
            serde_json::to_value(install_request)?
        } else {
            bail!("Either path or url must be provided");
        };

        let install_url = if is_dev_installation {
            multiaddr_to_url(multiaddr, "admin-api/dev/install-dev-application")?
        } else {
            multiaddr_to_url(multiaddr, "admin-api/dev/install-application")?
        };

        let install_response = get_response(
            &client,
            install_url,
            Some(install_request),
            &config.identity,
            POST,
        )
        .await?;

        if !install_response.status().is_success() {
            let status = install_response.status();
            let error_text = install_response.text().await?;
            bail!(
                "Application installation failed with status: {}. Error: {}",
                status,
                error_text
            )
        }

        let body = install_response
            .json::<InstallApplicationResponse>()
            .await?;

        info!(
            "Application installed successfully. Application ID: {}",
            body.data.application_id
        );

        Ok(())
    }
}
