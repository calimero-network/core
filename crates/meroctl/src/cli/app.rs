use clap::{Parser, Subcommand};
use eyre::Result as EyreResult;

use super::RootArgs;
use crate::cli::app::get::GetCommand;
use crate::cli::app::install::InstallCommand;
use crate::cli::app::list::ListCommand;

mod get;
mod install;
mod list;

#[derive(Debug, Parser)]
pub struct AppCommand {
    #[command(subcommand)]
    pub subcommand: AppSubCommands,
}

#[derive(Debug, Subcommand)]
pub enum AppSubCommands {
    Get(GetCommand),
    Install(InstallCommand),
    #[command(alias = "ls")]
    List(ListCommand),
}

impl AppCommand {
    pub async fn run(self, args: RootArgs) -> EyreResult<()> {
        match self.subcommand {
            AppSubCommands::Get(get) => get.run(args).await,
            AppSubCommands::Install(install) => install.run(args).await,
            AppSubCommands::List(list) => list.run(args).await,
        }
    }
}
