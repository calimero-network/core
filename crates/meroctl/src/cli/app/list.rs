use calimero_server_primitives::admin::ListApplicationsResponse;
use clap::Parser;
use eyre::{bail, Result as EyreResult};
use reqwest::Client;

use crate::cli::RootArgs;
use crate::common::RequestType::GET;
use crate::common::{get_response, multiaddr_to_url};
use crate::config_file::ConfigFile;

#[derive(Debug, Parser)]
pub struct ListCommand;

impl ListCommand {
    pub async fn run(self, root_args: RootArgs) -> EyreResult<()> {
        let path = root_args.home.join(&root_args.node_name);
        if !ConfigFile::exists(&path) {
            bail!("Config file does not exist")
        }
        let Ok(config) = ConfigFile::load(&path) else {
            bail!("Failed to load config file");
        };
        let Some(multiaddr) = config.network.server.listen.first() else {
            bail!("No address.")
        };

        let url = multiaddr_to_url(multiaddr, "admin-api/dev/applications")?;
        let client = Client::new();
        let response = get_response(&client, url, None::<()>, &config.identity, GET).await?;

        if !response.status().is_success() {
            bail!("Request failed with status: {}", response.status())
        }

        let api_response: ListApplicationsResponse = response.json().await?;
        let app_list = api_response.data.apps;

        #[allow(clippy::print_stdout)]
        for app in app_list {
            println!("{}", app.id);
        }

        Ok(())
    }
}
