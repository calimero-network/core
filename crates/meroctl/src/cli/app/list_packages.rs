use clap::Parser;
use eyre::Result;

use crate::cli::Environment;

#[derive(Copy, Clone, Parser, Debug)]
#[command(about = "List all packages")]
pub struct ListPackagesCommand;

impl ListPackagesCommand {
    pub async fn run(self, environment: &mut Environment) -> Result<()> {
        let client = environment.client()?;

        let response = client.list_packages().await?;

        environment.output.write(&response);
        Ok(())
    }
}
