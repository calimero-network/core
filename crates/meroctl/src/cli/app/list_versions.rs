use clap::Parser;
use eyre::Result;

use crate::cli::Environment;

#[derive(Clone, Parser, Debug)]
#[command(about = "List versions of a package")]
pub struct ListVersionsCommand {
    #[arg(help = "Package name (e.g., com.example.myapp)")]
    pub package: String,
}

impl ListVersionsCommand {
    pub async fn run(self, environment: &mut Environment) -> Result<()> {
        let client = environment.client()?;

        let response = client.list_versions(&self.package).await?;

        environment.output.write(&response);
        Ok(())
    }
}
