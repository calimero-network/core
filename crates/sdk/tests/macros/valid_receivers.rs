use calimero_sdk::app;

#[app::state]
struct MyType<'a>(&'a ());

#[app::logic]
impl<'a> MyType<'a> {
    // #[app::destroy]
    // pub fn method_00(self) {}
    pub fn method_01(&self) {}
    pub fn method_02(&mut self) {}
    // #[app::destroy]
    // pub fn method_03(self: Self) {}
    // #[app::destroy]
    // pub fn method_04(self: (Self)) {}
    // #[app::destroy]
    // pub fn method_05(mut self: Self) {}
    // #[app::destroy]
    // pub fn method_06(mut self: (Self)) {}
    // #[app::destroy]
    // pub fn method_07(self: &'a Self) {}
    // #[app::destroy]
    // pub fn method_08(self: &'a (Self)) {}
    pub fn method_09(mut self: &'a Self) {}
    pub fn method_10(mut self: &'a (Self)) {}
    pub fn method_11(self: &'a mut Self) {}
    pub fn method_12(self: &'a mut (Self)) {}
    pub fn method_13(mut self: &'a mut Self) {}
    pub fn method_14(mut self: &'a mut (Self)) {}
    // #[app::destroy]
    // pub fn method_15(self: MyType<'a>) {}
    // #[app::destroy]
    // pub fn method_16(self: (MyType<'a>)) {}
    // #[app::destroy]
    // pub fn method_17(mut self: MyType<'a>) {}
    // #[app::destroy]
    // pub fn method_18(mut self: (MyType<'a>)) {}
    pub fn method_19(self: &'a MyType<'a>) {}
    pub fn method_20(self: &'a (MyType<'a>)) {}
    pub fn method_21(mut self: &'a MyType<'a>) {}
    pub fn method_22(mut self: &'a (MyType<'a>)) {}
    pub fn method_23(self: &'a mut MyType<'a>) {}
    pub fn method_24(self: &'a mut (MyType<'a>)) {}
    pub fn method_25(mut self: &'a mut MyType<'a>) {}
    pub fn method_26(mut self: &'a mut (MyType<'a>)) {}
    // pub fn method_27(self: OtherType) {}
    // pub fn method_28(self: &'a OtherType) {}
}

struct OtherType;

fn main() {}
