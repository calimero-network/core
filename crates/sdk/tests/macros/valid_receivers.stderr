error: incompatible target architecture, no polyfill available, only wasm32 is supported.
 --> tests/macros/valid_receivers.rs:6:1
  |
6 | #[app::logic]
  | ^^^^^^^^^^^^^
  |
  = note: this error originates in the attribute macro `app::logic` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: unnecessary parentheses around type
  --> tests/macros/valid_receivers.rs:25:36
   |
25 |     pub fn method_10(mut self: &'a (Self)) {}
   |                                    ^    ^
   |
   = note: `#[warn(unused_parens)]` on by default
help: remove these parentheses
   |
25 -     pub fn method_10(mut self: &'a (Self)) {}
25 +     pub fn method_10(mut self: &'a Self) {}
   |

warning: unnecessary parentheses around type
  --> tests/macros/valid_receivers.rs:27:36
   |
27 |     pub fn method_12(self: &'a mut (Self)) {}
   |                                    ^    ^
   |
help: remove these parentheses
   |
27 -     pub fn method_12(self: &'a mut (Self)) {}
27 +     pub fn method_12(self: &'a mut Self) {}
   |

warning: unnecessary parentheses around type
  --> tests/macros/valid_receivers.rs:29:40
   |
29 |     pub fn method_14(mut self: &'a mut (Self)) {}
   |                                        ^    ^
   |
help: remove these parentheses
   |
29 -     pub fn method_14(mut self: &'a mut (Self)) {}
29 +     pub fn method_14(mut self: &'a mut Self) {}
   |

warning: unnecessary parentheses around type
  --> tests/macros/valid_receivers.rs:39:32
   |
39 |     pub fn method_20(self: &'a (MyType<'a>)) {}
   |                                ^          ^
   |
help: remove these parentheses
   |
39 -     pub fn method_20(self: &'a (MyType<'a>)) {}
39 +     pub fn method_20(self: &'a MyType<'a>) {}
   |

warning: unnecessary parentheses around type
  --> tests/macros/valid_receivers.rs:41:36
   |
41 |     pub fn method_22(mut self: &'a (MyType<'a>)) {}
   |                                    ^          ^
   |
help: remove these parentheses
   |
41 -     pub fn method_22(mut self: &'a (MyType<'a>)) {}
41 +     pub fn method_22(mut self: &'a MyType<'a>) {}
   |

warning: unnecessary parentheses around type
  --> tests/macros/valid_receivers.rs:43:36
   |
43 |     pub fn method_24(self: &'a mut (MyType<'a>)) {}
   |                                    ^          ^
   |
help: remove these parentheses
   |
43 -     pub fn method_24(self: &'a mut (MyType<'a>)) {}
43 +     pub fn method_24(self: &'a mut MyType<'a>) {}
   |

warning: unnecessary parentheses around type
  --> tests/macros/valid_receivers.rs:45:40
   |
45 |     pub fn method_26(mut self: &'a mut (MyType<'a>)) {}
   |                                        ^          ^
   |
help: remove these parentheses
   |
45 -     pub fn method_26(mut self: &'a mut (MyType<'a>)) {}
45 +     pub fn method_26(mut self: &'a mut MyType<'a>) {}
   |
